yt_dlp/extractor/redgifs.py

```diff
@@ -76,10 +80,22 @@ def _call_api(self, ep, video_id, *args, **kwargs):
             self._fetch_oauth_token(video_id)
         assert 'authorization' in self._API_HEADERS
 
-        headers = dict(self._API_HEADERS)
-        headers['x-customheader'] = f'https://www.redgifs.com/watch/{video_id}'
-        data = self._download_json(
-            f'https://api.redgifs.com/v2/{ep}', video_id, headers=headers, *args, **kwargs)
+        for _ in range(MAX_TOKEN_REFRESH_ATTEMPTS + 1):
+            try:
+                headers = dict(self._API_HEADERS)
+                headers['x-customheader'] = f'https://www.redgifs.com/watch/{video_id}'
+
+                data = self._download_json(
+                    f'https://api.redgifs.com/v2/{ep}', video_id, headers=headers, *args, **kwargs)
+                break
+            except ExtractorError as e:
+                if not isinstance(e.cause, urllib.error.HTTPError) or e.cause.code != 401:
+                    raise
+
+                # Likely the temporary token has expired
+                del self._API_HEADERS['authorization']
+                self._fetch_oauth_token(video_id)
+
```

Comment:
- the first `_fetch_oauth_token` call can be moved inside the loop to get rid of the duplication
 - the `assert` line can also be removed since `_fetch_oauth_token` is fatal
 - `data` should be defined before `try/except` in case there are two consecutive 401 responses for some unforeseen reason
```py
    def _call_api(self, ep, video_id, *args, **kwargs):
        data = {'error': '401 Unauthorized'}  # in case of two 401 responses
        for _ in range(2):
            if 'authorization' not in self._API_HEADERS:
                self._fetch_oauth_token(video_id)
            try:
                headers = dict(self._API_HEADERS)
                headers['x-customheader'] = f'https://www.redgifs.com/watch/{video_id}'
                data = self._download_json(
                    f'https://api.redgifs.com/v2/{ep}', video_id, headers=headers, *args, **kwargs)
                break
            except ExtractorError as e:
                if not isinstance(e.cause, urllib.error.HTTPError) or e.cause.code != 401:
                    raise
                del self._API_HEADERS['authorization']  # refresh the token

        if 'error' in data:
            raise ExtractorError(f'RedGifs said: {data["error"]}', video_id=video_id)
        return data
```

Reply:
Thanks, great feedback. I was a bit hasty I suppose, especially with the case where it failed on the second attempt.

I decided not to define data before but actually to re-raise on the second attempt from the except handler. That way the original exception bubbles up.